test/functional/components/typography/body.fixture.html

```diff
@@ -0,0 +1,11 @@
+<!DOCTYPE html>
+<html lang="en">
+  <head>
+    <meta charset="UTF-8" />
+    <title>Component View</title>
+  </head>
+  <body>
+    <div id="root"></div>
+    <script src="body.fixture.jsx"></script>
+  </body>
+</html>
```

Comment:
Is it possible to include jsx in the test file? It seems like a lot of work to create a .fixture.html, .fixture.jsx, and .spec.js file for every component we want to test. For example, you could save duplication by producing a body component instance within the test as you already have an array of test cases with all the sizes. Do you see the .fixture.html file always being the same except loading a different jsx file?

Reply:
I considered this in my approach and the answer is yes, you can use JSX inside TestCafe specs, but with limitations. For example, if the JSX in the spec file has `<MyCustomComponent>`, then I need to import `MyCustomComponent` into the spec file and rely on TestCafe's transpilation of `MyCustomComponent`. However, TestCafe's transpilation doesn't use the babel configuration that we've set up for our project, so it ends up choking on some files, like our Stylus files. This predicament is explained pretty well by [this comment](https://github.com/DevExpress/testcafe/issues/2659#issuecomment-492436113) (if that link doesn't take you to the comment, look for the comment by anonmos).

Even if TestCafe were to be able to transpile our JSX properly, that would tranpile the JSX within the _spec_ file, which runs in Node. We need it to transpile it in such a way that it can actually render react elements on the page.  Putting the JSX in a `ClientFunction` would work if we loaded React and ReactDOM onto the fixture HTML page, but we still wouldn't be able to use our custom components within the JSX because we'd still be importing the custom components within our spec file, when we really need the custom component code on the fixture page itself (and we run into the transpilation problem I just described anyway). We'd only be able to render regular DOM elements, because those are provided by ReactDOM, which would be on the fixture page. Hopefully that makes sense.

Your questioning made me think through this problem more though. I've added a commit that solves the issue in a different way. It doesn't allow us to use JSX with custom components in our spec file, but I don't think that's the important part anyway. I think the important part is to have the rendering logic for each test case inside the spec file instead of the fixture.

In the commit I added, we use a single fixture HTML file. The fixture HTML file "imports" a single JSX file which imports all our custom components we want to test. It also adds a global method on window (`window.renderSerializedReactElement`). In our spec file, _outside_ of any ClientFunction, we can call `React.createElement("MyCustomComponent", props, children`)`. Now, when using `React.createElement`, you typically pass the actual component function as the first argument. In our case, we're passing a string, which is important. When the test runs, React.createElement will run, which spits out a plain JavaScript object representing the element graph that needs to be rendered, but it retains the string `MyCustomComponent` for the type (remember, we can't import the `MyCustomComponent` component inside the spec file). In our case, we need to render the element graph on the _page_, not in the _spec file_, so we send that element graph object to the page using a `ClientFunction` we've defined. The `ClientFunction` then, on the fixture page, passes the element graph to `window.renderSerializedReactElement` (which our fixture has provided). `window.renderSerializedReactElement` then "deserializes" the object and swaps out the `MyCustomComponent` string (which is the value for the `type` property on the element graph object) with the actual `MyCustomComponent` component function. We then feed that deserialized element graph to `React.render`, which then renders the content to the page.

The downside about this approach is that when we want to test a new component, we need to import that component in our fixture file. I added a check in the fixture file that should help us detect when we fail to do so.

Anyway, take a look and let me know what you think. I do like it more than what I had.

Reply:
Alternatively, we could just use a different test framework. 😆